server/asset/eth/cache.go

```diff
@@ -7,130 +7,107 @@
-		orphaned: orphaned,
-	}
-	cache.blocks[hash] = blk
-
-	if !orphaned {
-		cache.mainchain[height] = blk
-		if height > cache.best.height {
-			cache.best.height = height
-			cache.best.hash = hash
+// poll pulls the best hash from an eth node and compares that to a stored
+// hash. If the same does nothing. If different, updates the stored hash and
+// notifies listeners on block chans.
+func (hc *hashCache) poll(ctx context.Context) {
+	send := func(reorg bool, err error) {
+		if err != nil {
+			hc.log.Error(err)
 		}
+		hc.signalMtx.Lock()
+		for c := range hc.blockChans {
+			select {
+			case c <- &asset.BlockUpdate{
+				Reorg: reorg,
+				Err:   err,
+			}:
+			default:
+				hc.log.Error("failed to send block update on blocking channel")
+			}
+		}
+		hc.signalMtx.Unlock()
 	}
-	return blk, nil
-}
-
-// Get the best known block height for the blockCache.
-func (cache *blockCache) tipHeight() uint64 {
-	cache.mtx.Lock()
-	defer cache.mtx.Unlock()
-	return cache.best.height
-}
-
-// Get the best known block hash in the blockCache.
-func (cache *blockCache) tipHash() common.Hash {
-	cache.mtx.RLock()
-	defer cache.mtx.RUnlock()
-	return cache.best.hash
-}
-
-// Get the best known block height in the blockCache.
-func (cache *blockCache) tip() ethBlock {
-	cache.mtx.RLock()
-	defer cache.mtx.RUnlock()
-	return cache.best
-}
-
-// Trigger a reorg, setting any blocks at or above the provided height as
-// orphaned and removing them from mainchain, but not the blocks map. reorg
-// clears the best block, so should always be followed with the addition of a
-// new mainchain block.
-func (cache *blockCache) reorg(from uint64) {
-	if from < 0 {
+	hc.mtx.Lock()
+	defer hc.mtx.Unlock()
+	bhdr, err := hc.node.bestHeader(ctx)
+	if err != nil {
+		send(false, fmt.Errorf("error getting best block header from geth: %w", err))
 		return
 	}
-	cache.mtx.Lock()
-	defer cache.mtx.Unlock()
-	for height := from; height <= cache.best.height; height++ {
-		block, found := cache.mainchain[height]
-		if !found {
-			cache.log.Errorf("reorg block not found on mainchain at height %d for a reorg from %d to %d", height, from, cache.best.height)
-			continue
-		}
-		// Delete the block from mainchain.
-		delete(cache.mainchain, block.height)
-		// Store an orphaned block in the blocks cache.
-		cache.blocks[block.hash] = &ethBlock{
-			hash:     block.hash,
-			height:   block.height,
-			orphaned: true,
+	if bhdr.Hash() == hc.best.hash {
+		// Same hash, nothing to do.
+		return
+	}
+	update := func(reorg bool, fastBlocks bool) {
+		hash := bhdr.Hash()
+		height := bhdr.Number.Uint64()
+		str := fmt.Sprintf("Tip change from %s (%d) to %s (%d).",
+			hc.best.hash, hc.best.height, hash, height)
+		switch {
+		case reorg:
+			str += " Detected reorg."
+		case fastBlocks:
+			str += " Fast blocks."
 		}
+		hc.log.Info(str)
+		hc.best.hash = hash
+		hc.best.height = height
+	}
+	if bhdr.ParentHash == hc.best.hash {
+		// Sequencial hash, report a block update.
```

Comment:
Sequential. Here and one place in tests.

Reply:
Fixed.